### ISSUE TITLE
global variables erased by each loop when using same variable name

### ISSUE THREAD

**Issue by SergioCrisostomo** [pugjs/pug#2365]
If I do:

    app.get('/', (req, res) => {
      var data = {title: 'Hej!', foo: 'bar'};
      res.render('index', data);
    });

and in my `index.pug` I have

	doctype html
	html(lang="en")
	  head
	    title= title
	   
	  body
	    h1= title
	    p= foo

it works as expected and I get text in both `h1` and `p`.

But if I have a loop in the `.pug` file like

	doctype html
	html(lang="en")
	  head
	    title= title
	   
	  body
	    h1= title
	    p= foo
	    each foo in [1, 2, 3]
	        h3= foo

then the `foo` variable in the global scope gets deleted/empty and only the scoped `foo` works. [This worked in Jade](http://jsfiddle.net/zkukg7fe/), but not in Pug. Made a test case [here (link)](https://ide.c9.io/sergiocrisostomo/pug-loop-bug). 

So basically the `each` loop creates a new block scope for `foo`, but the global `foo` gets erased.

Is this a breaking change from Jade to Pug? or a bug?

**Comment by niftylettuce** [comment]
the same bug I have!  it must be a breaking change from switching from jade to pug, because I never had this before pug?

**Comment by niftylettuce** [comment]
this is a critical bug please fix ASAP @ForbesLindesay, not sure where this is...

**Comment by SergioCrisostomo** [comment]
@niftylettuce yes, its a breaking change. This is complex to fix, so I do not think this will be fixed soon.

Renaming the vars making sure you don't repeat them was the solution for me.

**Comment by niftylettuce** [comment]
We should mention this here https://github.com/pugjs/pug/issues/2305

**Comment by niftylettuce** [comment]
So we can do this in the meanwhile to continue using pug:

```js
npm i --save SergioCrisostomo/pug-code-gen#use-let
```

**Comment by feross** [comment]
This issue should really be mentioned here: https://github.com/pugjs/pug/issues/2305 This bug makes migrating a large codebase to `pug` from `jade` substantially more risky and more work.

@ForbesLindesay Is there any reason that the fix here (https://github.com/SergioCrisostomo/pug-code-gen/commit/c4ad9556c2a5ae28b8e646fd5cb4afcf185b3aaf) can't be applied to `pug`?

**Comment by TimothyGu** [comment]
@feross, it was an intentional change. See #1983 and pugjs/pug-code-gen#1.

**Comment by feross** [comment]
@TimothyGu Has anyone actually tried using `pug` on a reasonably large codebase? This change makes it incredibly to introduce subtle bugs. Variables that look like they're local really ought to be local. Is there really no way to give them a random prefix so they don't conflict with variables that the user passes into `render()`?

**Comment by TimothyGu** [comment]
Update: this is fixed by pugjs/pug-code-gen#28, which is pending review.

**Comment by niftylettuce** [comment]
Hey @TimothyGu @feross @SergioCrisostomo @ForbesLindesay this is a bug again - legacy support was removed when @SergioCrisostomo rid of his fork which had the let support `npm i --save SergioCrisostomo/pug-code-gen#use-let` -- and the latest version doesn't have this.

**Comment by SergioCrisostomo** [comment]
@niftylettuce I wasn't aware my fork was being used. Sorry for that.
I re-forked and uploaded the branch again. I think this will help, and I will now remember to keep it there :)

**Comment by niftylettuce** [comment]
Thanks @SergioCrisostomo - I think I got super lucky 🍀 :crossed_fingers: and only had to fix one of my views for this issue - so far at least!  I just changed variable name used outside of loop to `_prefixedWithUnderscore`.
